api/registry/registry.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package registry
+
+import (
+	"errors"
+	"sync"
+	"sync/atomic"
+
+	"github.com/open-telemetry/opentelemetry-go/api/unit"
+)
+
+type Sequence uint64
+
+type Option func(Variable) Variable
+
+type Variable struct {
+	Name        string
+	Description string
+	Unit        unit.Unit
+	Type        Type
+	Sequence    Sequence // 0 == unregistered
+	Status      error    // Indicates registry conflict
+}
```

Comment:
Variable is used for Metric, Measure, Keys (Keys are used in Metrics, and Trace Attributes). This generalization is probably more confusing.  Also, when a new variable is created it is registered in nameToMap irrespective of where it is used. For example, if Variable used for Trace Attribute, it will get registered. What is the use of registering it? Registration is more applicable to Metrics.

Also, if the same name is used for Attribute in trace and for Metric, the Status of Variable for one of them will result into ErrConflictingDef? What is the use of this Status.

This PR is simply restructuring the code so may be I can open an issue for this.

Reply:
Thank you! I appreciate your willingness to discuss this in a follow-on issue. 

You pin-pointed one of the negative aspects of unifying the treatment of Keys throughout the code. If there are reasons to pre-register keys and metrics, there may be reasons not to do so as well. 

As for benefits in having uniform structure: means that span attributes and log fields can have units and description metadata. The benefits of registering keys is separate. We can efficiently index based on registered keys, for metrics pre-aggregation, but we can also avoid repeatedly encoding registered keys into an output format. 

I definitely want to support _unregistered_ keys, to allow for certain proxies and applications that generate span/log keys at runtime. I might add a method to generate these `key.Unregistered("blah")`, but I'm not sure what the best name for it is. I left a comment saying that a Key w/ `Sequence == 0` is unregistered, to help address this case, but didn't create a method for generating unregistered keys.

You're also right to point out some confusion over the Status error. I like to think that it's not a hard error to use conflicting names in an application, since we commonly link in third-party code that we do not control. In that sense, the registry as it stands here simply maintains separate registrations for differently-typed variables with the same name. I.e., the uniqueness is over (name, type), not just the name. I think this is all relatively unimportant since real observability systems need to deal with data from multiple versions of an application, where conflicts and multiply-defined types are certain to happen.